### PR TITLE
Feat/reduce plugins scope

### PR DESCRIPTION
**Pull request by rsanteri**
### Requirements for a pull request

- [ ] Unit tests related to the change have been updated 
- [ ] Documentation related to the change has been updated 

### Description of the Change

This PR reduces scope of the plugins systems to make it easier to maintain in the future.

Removes slots:
- run-tab
- headless

Removes exposed data
- task
- run
- appinfo

This PR also adds some feature to plugin system. Plugin slot will now provide available IDs from current URL to every plugin. Knowing flow_id, run_number or task_id might be useful for the plugin.

Other change this brings is that now plugins needs to define configurations in manifest.json instead of register function. register is replaced with onReady. It is recommended to use onReady to wrap plugin code but not absolute necessity.

### Possible Drawbacks

Existing plugins using these slots or data must be updated to work with new API.

### CODE REVIEW

File: plugin-api/Examples/hello-world/plugin.html [hunk 1]
```diff
@@ -19,13 +19,11 @@
...
       // plugin is placed to correct spot in UI.
-      Metaflow.register({ slot: 'run-header', container: 'titled-container'}, () => {
+      Metaflow.register({ slot: 'run-header', container: 'titled-container'}, (config, resource) => {
         // Set fixed height for plugin
         Metaflow.setHeight(40);
-        // Subscribe to any update on run data and show run number in plugin.
-        Metaflow.subscribe(['run'], (event) => {
-          document.getElementById('container').innerHTML = `Hello World! Run number is ${event.data.run_number}`;
-        });
+        // Show run id from URL.
```
Comment (obgibson) [inline plugin-api/Examples/hello-world/plugin.html:25]: You don't need to subscribe to "metadata"?

File: plugin-api/MetaflowPluginAPI.js [hunk 2]
```diff
@@ -1,37 +1,37 @@
 const VERSION_INFO = {
-  api: '0.13.0'
-}
+  api: '0.13.3',
```
Comment (obgibson) [inline plugin-api/MetaflowPluginAPI.js:2]: As this is a breaking change, shouldn't we bump it more than a patch version or two?

File: plugin-api/MetaflowPluginAPI.js [hunk 3]
```diff
@@ -73,73 +73,93 @@ const Metaflow = {
...
+   * @param {(event: { path: string, data: * }) => void} fn
    */
   subscribe(paths, fn) {
     Listeners.push(fn);
-    window.parent.postMessage({ name: window.name, type: 'PluginSubscribeToData', paths: paths }, '*')
+    window.parent.postMessage({ name: window.name, type: 'PluginSubscribeToData', paths: paths }, '*');
   },
   /**
-   * Subsribe to events 
+   * Subsribe to events
```
Comment (obgibson) [inline plugin-api/MetaflowPluginAPI.js:98]: typo - `Subscribe`

File: src/components/Plugins/PluginManager.tsx [hunk 4]
```diff
@@ -49,11 +49,11 @@ type PluginVersionInfo = {
 // Constants. Plugin will not render if it doesn't satisfy SUPPORTED_PLUGIN_API_VERSION.
 //
 
-export type AllowedSlot = 'run-header' | 'run-tab' | 'task-details' | 'headless';
+export type AllowedSlot = 'task-details' | 'run-header';
 
 const SUPPORTED_PLUGIN_API_VERSION = '0.13.0';
-const RECOMMENDED_PLUGIN_API_VERSION = '0.13.0';
-const ALLOWED_SLOTS: AllowedSlot[] = ['run-header', 'run-tab', 'task-details', 'headless'];
+const RECOMMENDED_PLUGIN_API_VERSION = '0.13.1';
```
Comment (obgibson) [inline src/components/Plugins/PluginManager.tsx:55]: This is 0.13.3 up above

File: plugin-api/MetaflowPluginAPI.js [hunk 5]
```diff
@@ -73,73 +73,93 @@ const Metaflow = {
...
-}
 
-if (typeof exports !== "undefined") {
-  exports.Metaflow = Metaflow
+  subscribeToMetadata(fn) {
+    this.subscribe(['metadata'], fn);
+  },
+
+  subscribeToRunMetadata(fn) {
+    this.subscribe(['run-metadata'], fn);
```
Comment (obgibson) [inline plugin-api/MetaflowPluginAPI.js:155]: I don't see `run-metadata` in the documentation anywhere. I think we should have it listed with `metadata` somewhere so plugin writers don't have to consult to code to find it.

**Review by obgibson — CHANGES_REQUESTED** [review]

File: plugin-api/Examples/hello-world/plugin.html (hunk 1 above)
Comment (rsanteri) [reply plugin-api/Examples/hello-world/plugin.html:24]: Removed subscribe due not being needed in this example anymore

File: plugin-api/MetaflowPluginAPI.js (hunk 2 above)
Comment (rsanteri) [reply plugin-api/MetaflowPluginAPI.js]: Upped

**Review by obgibson — APPROVED** [review]

File: plugin-api/MetaflowPluginAPI.js [hunk 6]
```diff
@@ -1,145 +1,193 @@
...
+      "*"
+    );
   },
-}
+
+  subscribeToMetadata(fn) {
+    this.subscribe(["metadata"], fn);
+  },
+
+  subscribeToRunMetadata() {
```
Comment (obgibson) [inline plugin-api/MetaflowPluginAPI.js:182]: This signature needs `fn`

**Review by obgibson — CHANGES_REQUESTED** [review]

File: plugin-api/MetaflowPluginAPI.js (hunk 6 above)
Comment (rsanteri) [reply plugin-api/MetaflowPluginAPI.js]: Added

File: docs/plugin-system.md [hunk 7]
```diff
@@ -36,13 +36,13 @@ A basic plugin could be something like
     <script src="MetaflowPluginAPI.js"></script>
     <script>
       (function () {
-        Metaflow.register('run-header', () => {
+        Metaflow.onReady((configuration, resource) => {
```
Comment (skirdey) [inline docs/plugin-system.md:39]: below it mentions `There are two implemented plugin slots, `run-header` and `task-details`. The desired slot must be given as parameter to a plugin API register message.` but here it removes run-header registry call - is there some equivalent of it?

File: docs/plugin-system.md [hunk 8]
```diff
@@ -59,24 +59,16 @@ This plugin is registered to be rendered in the run header section (path /FLOW_I
 
 ## Plugin slots
 
-There are three implemented plugin slots. `run-header`, `task-details`, and `headless`. The desired slot must be given as a parameter to a plugin API register message.
+There are two implemented plugin slots, `run-header` and `task-details`. The desired slot must be given as parameter to a plugin API register message.
```
Comment (skirdey) [inline docs/plugin-system.md:62]: it should probably mention plugin-api/Examples/hello-world/manifest.json and the definition of the slot for clarity

File: plugin-api/Examples/react-plugin/app.jsx [hunk 9]
```diff
@@ -4,22 +4,22 @@ import { Metaflow } from '../../MetaflowPluginAPI'
...
 
-  // subscribe to listen run resource.
+  // subscribe to listen metadata resource.
   useEffect(() => {
-    Metaflow.subscribe(['run'], (event) => {
-      setRun(event.data);
+    Metaflow.subscribeToMetadata((event) => {
+      setMetadata(event.data);
     });
-    Metaflow.setHeight(100)
```
Comment (skirdey) [inline plugin-api/Examples/react-plugin/app.jsx:14]: trying to understand here why setHeight(100) was replaced by setHeight()?

File: plugin-api/MetaflowPluginAPI.js [hunk 10]
```diff
@@ -1,145 +1,193 @@
 const VERSION_INFO = {
-  api: '0.13.0'
-}
+  api: "0.14.0",
```
Comment (skirdey) [inline plugin-api/MetaflowPluginAPI.js:2]: this seems to be a breaking change update, can we do a semantic version?

**Review by skirdey — CHANGES_REQUESTED** [review]
very few minor things

**Review by obgibson — COMMENTED** [review]
Tests workforme

File: docs/plugin-system.md (hunk 7 above)
Comment (rsanteri) [reply docs/plugin-system.md:39]: Updated to have mention about manifest.json

File: plugin-api/Examples/react-plugin/app.jsx (hunk 9 above)
Comment (rsanteri) [reply plugin-api/Examples/react-plugin/app.jsx:14]: setHeight has optional fixedHeight parameter. When no parameter is given, API will try to find out the full height of document and use that. There wasn't really need for fixed height here.

File: plugin-api/MetaflowPluginAPI.js (hunk 10 above)
Comment (rsanteri) [reply plugin-api/MetaflowPluginAPI.js:2]: Yes i imagine we should up it to 1.0.0

Comment (skirdey) [reply plugin-api/MetaflowPluginAPI.js]: lets do it and I think we can merge 1.0.0

**Pull request merged**